http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/OneTimeReadTimeoutHandler.java

```diff
@@ -16,34 +16,24 @@
 package software.amazon.awssdk.http.nio.netty.internal;
 
 import io.netty.channel.ChannelHandlerContext;
-import io.netty.channel.SimpleChannelInboundHandler;
 import io.netty.handler.timeout.ReadTimeoutHandler;
-import io.netty.util.ReferenceCountUtil;
+import java.time.Duration;
 import java.util.concurrent.TimeUnit;
 import software.amazon.awssdk.annotations.SdkInternalApi;
 
 /**
- * Handler to add an one-time {@link ReadTimeoutHandler} to the pipeline and remove it afterwards.
+ * A one-time read timeout handler that removes itself from the pipeline after
+ * the next successful read.
  */
 @SdkInternalApi
-public final class OneTimeReadTimeoutHandler extends SimpleChannelInboundHandler {
-
-    private static final String READ_TIMEOUT_HANDLER_NAME = "RemoveAfterReadTimeoutHandler";
-    private final long readTimeoutMillis;
-    private final TimeUnit timeUnit;
-
-    OneTimeReadTimeoutHandler(long readTimeoutMillis, TimeUnit timeUnit) {
-        this.readTimeoutMillis = readTimeoutMillis;
-        this.timeUnit = timeUnit;
+public final class OneTimeReadTimeoutHandler extends ReadTimeoutHandler {
+    OneTimeReadTimeoutHandler(Duration timeout) {
+        super(timeout.toMillis(), TimeUnit.MILLISECONDS);
     }
 
     @Override
-    public void channelRead0(ChannelHandlerContext ctx, Object msg) {
-        ReferenceCountUtil.retain(msg);
-        ctx.pipeline().addFirst(READ_TIMEOUT_HANDLER_NAME, new ReadTimeoutHandler(readTimeoutMillis, timeUnit));
-        ctx.fireChannelRead(msg);
-
-        ctx.pipeline().remove(READ_TIMEOUT_HANDLER_NAME);
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
         ctx.pipeline().remove(this);
+        super.channelRead(ctx, msg);
```

Comment:
Shouldn't it be `channelRead` first and then `remove`?

Reply:
I don't think so. If a downstream handler removes us first (for example if the channel is released and `HandlerRemovingChannelPool` removes it), then `remove` will fail; so we remove the read timeout handler first, then fire the read down the rest of the pipeline.

Reply:
Oh I see, yeah.

Is there any reason why `super.channelRead` is being used? Should we do `ctx.fireChannelRead(msg)` ?

Reply:
Yeah that would work, but technically this is just a specialized `ReadTimeoutHandler` so I think it makes more sense to just defer to it's `channelRead` impl.

Reply:
Looking at the source code of `IdleStateHandler`, there's some extra logic there. Probably doesn't matter for now, but I'd lean towards to using `ctx.fireChannelRead` directly just in case there's some broken change introduced in the future(somehow).

```java
    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
        if (readerIdleTimeNanos > 0 || allIdleTimeNanos > 0) {
            reading = true;
            firstReaderIdleEvent = firstAllIdleEvent = true;
        }
        ctx.fireChannelRead(msg);
    }

```

Reply:
I think that's more my point though. Not calling `super.channelRead` since there's extra logic there could also break behavior no?

Reply:
Hmm, yeah, I guess that makes sense.